flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/optimize/StreamNonDeterministicUpdatePlanVisitor.java

```diff
@@ -0,0 +1,1009 @@
+                if (ndCall.isPresent()) {
+                    // all columns from table function scan cannot satisfy the required determinism
+                    List<Integer> unsatisfiedColumns =
+                            requireDeterminism.toList().stream()
+                                    .filter(index -> index >= leftFieldCnt)
+                                    .collect(Collectors.toList());
+                    if (!unsatisfiedColumns.isEmpty()) {
+                        String errorMsg =
+                                generateNonDeterministicColumnsErrorMessage(
+                                        unsatisfiedColumns,
+                                        correlate.getRowType(),
+                                        correlate,
+                                        null,
+                                        ndCall);
+                        throw new TableException(errorMsg);
+                    }
+                }
+                // evaluate required determinism from input
+                List<Integer> fromLeft =
+                        requireDeterminism.toList().stream()
+                                .filter(index -> index < leftFieldCnt)
+                                .collect(Collectors.toList());
+                if (fromLeft.isEmpty()) {
+                    return transmitDeterminismRequirement(correlate, NO_REQUIRED_DETERMINISM);
+                }
+                return transmitDeterminismRequirement(correlate, ImmutableBitSet.of(fromLeft));
+            }
+
+        } else if (rel instanceof StreamPhysicalLookupJoin) {
+            if (inputInsertOnly(rel) || requireDeterminism.isEmpty()) {
+                return transmitDeterminismRequirement(rel, NO_REQUIRED_DETERMINISM);
+            } else {
+                /**
+                 * if input has updates, the lookup join may produce non-deterministic result itself
+                 * due to backed lookup source which data may change over time, we can try to
+                 * eliminate this non-determinism by adding materialization to the join operator,
+                 * but still exists non-determinism we cannot solve: 1. join condition 2. the inner
+                 * calc in lookJoin.
+                 */
+                StreamPhysicalLookupJoin lookupJoin = (StreamPhysicalLookupJoin) rel;
+
+                // required determinism cannot be satisfied even upsert materialize was enabled if:
+                // 1. remaining join condition contains non-deterministic call
+                if (lookupJoin.remainingCondition().isDefined()) {
+                    RexNode rexNode = lookupJoin.remainingCondition().get();
+                    checkNonDeterministicCondition(rexNode, lookupJoin);
+                }
+                // 2. inner calc in lookJoin contains either non-deterministic condition or calls
+                if (lookupJoin.calcOnTemporalTable().isDefined()) {
```

Comment:
nit: we can use `JavaScalaConversionUtil.toJava(lookupJoin.calcOnTemporalTable())
                        .ifPresent(
                                c ->
                                        checkNonDeterministicRexProgram(
                                                requireDeterminism, c, lookupJoin));`
to avoid if branch